sdk/python/kfp/components/structures.py

```diff
@@ -74,60 +77,126 @@ class OutputSpec(base_model.BaseModel):
     description: Optional[str] = None
 
 
-class InputValuePlaceholder(base_model.BaseModel):
+class PlaceholderSerializationMixin(abc.ABC):
+    _FROM_PLACEHOLDER_REGEX: Union[str, type(NotImplemented)] = NotImplemented
+    _TO_PLACEHOLDER_TEMPLATE_STRING: Union[
+        str, type(NotImplemented)] = NotImplemented
+
+    @classmethod
+    def _is_input_placeholder(cls) -> bool:
+        field_names = {field.name for field in dataclasses.fields(cls)}
+        return "input_name" in field_names
+
+    @classmethod
+    def is_match(cls, placeholder: str) -> bool:
+        return re.match(cls._FROM_PLACEHOLDER_REGEX, placeholder) is not None
+
+    @classmethod
+    def from_placeholder(cls, placeholder: str) -> Any:
+        if cls._FROM_PLACEHOLDER_REGEX == NotImplemented:
+            raise NotImplementedError(
+                f'{cls.__name__} does not support placeholder parsing.')
+
+        matches = re.search(cls._FROM_PLACEHOLDER_REGEX, placeholder)
+        if matches is None:
+            raise ValueError(
+                f'Could not parse placeholder: {placeholder} into {cls.__name__}'
+            )
+        if cls._is_input_placeholder():
+            return cls(input_name=matches[1])
+        else:
+            return cls(output_name=matches[1])
+
+    def to_placeholder(self) -> str:
+        if self._TO_PLACEHOLDER_TEMPLATE_STRING == NotImplemented:
+            raise NotImplementedError(
+                f'{self.__class__.__name__} does not support creating placeholder strings.'
+            )
+        attr_name = 'input_name' if self._is_input_placeholder(
+        ) else 'output_name'
+        value = getattr(self, attr_name)
+        return self._TO_PLACEHOLDER_TEMPLATE_STRING.format(value)
+
+
+class InputValuePlaceholder(base_model.BaseModel,
+                            PlaceholderSerializationMixin):
     """Class that holds input value for conditional cases.
 
     Attributes:
         input_name: name of the input.
     """
     input_name: str
     _aliases = {'input_name': 'inputValue'}
+    _TO_PLACEHOLDER_TEMPLATE_STRING = "{{{{$.inputs.parameters['{}']}}}}"
+    _FROM_PLACEHOLDER_REGEX = r"\{\{\$\.inputs\.parameters\[(?:''|'|\")(.+?)(?:''|'|\")]\}\}"
 
 
-class InputPathPlaceholder(base_model.BaseModel):
+class InputPathPlaceholder(base_model.BaseModel, PlaceholderSerializationMixin):
```

Comment:
Do we still need `BaseModel`?

Reply:
I believe so, for backward compatibility purposes. `BaseModel` allows us to read inputs and output placeholders from V1 YAML JSON directly into the `*Placeholder` objects, without requiring any special handling.

Put differently, `BaseModel`s know how to handle deserializing data into primitives and into other `BaseModel`s (only). If we remove `BaseModel` as a parent of `*Placeholders`, the `ContainerSpec` object which contains `*Placeholders` will not know how to treat it.

Reply:
Sorry, I misread, I thought this was the BaseModel class we used with Pydantic.